packages/@aws-cdk/aws-cloudwatch/lib/dashboard.ts

```diff
@@ -127,7 +127,29 @@ export class Dashboard extends Resource {
       return;
     }
 
+    const warnings = allWidgetsDeep(widgets).flatMap(w => w.warnings ?? []);
+    for (const w of warnings) {
+      Annotations.of(this).addWarning(w);
+    }
+
     const w = widgets.length > 1 ? new Row(...widgets) : widgets[0];
     this.rows.push(w);
   }
 }
+
+function allWidgetsDeep(ws: IWidget[]) {
+  const ret = new Array<IWidget>();
+  ws.forEach(recurse);
+  return ret;
+
+  function recurse(w: IWidget) {
+    ret.push(w);
+    if (hasSubWidgets(w)) {
+      w.widgets.forEach(recurse);
+    }
+  }
```

Comment:
Is there a chance that we have a circular dependency in any `widgets`? That is, is it possible for a widget to be a subwidget of itself? If so we should add a depth check and throw a useful error message if we catch one.

Reply:
You can write code to do that, but it already won't work today with a stack blowing error. I'm not too concerned that someone would do it in practice and then not know what's going on.